rust/ql/lib/codeql/rust/internal/typeinference/FunctionType.qll

```diff
@@ -3,7 +3,28 @@ private import codeql.rust.internal.TypeInference
 private import codeql.rust.internal.PathResolution
 private import codeql.rust.internal.Type
 private import codeql.rust.internal.TypeMention
-private import codeql.rust.elements.Call
+
+private newtype TArgumentPosition =
+  TPositionalArgumentPosition(int i) {
+    i in [0 .. max([any(ParamList l).getNumberOfParams(), any(ArgList l).getNumberOfArgs()]) - 1]
+  } or
+  TSelfArgumentPosition()
+
+/** An argument position in a call. */
+class ArgumentPosition extends TArgumentPosition {
```

Comment:
I don't think `ArgumentPositon` should be here as it has utility beyond type inference. For instance, we can use it to simplify `TParameterPosition` in data flow. It might also be useful to add an `Expr getArgument(ArgumentPositon pos)` predicate to `Call` or `MethodCall`.